addons/l10n_ke_edi_tremol/static/src/components/ke_proxy_hook.js

```diff
@@ -0,0 +1,88 @@
+     * details, from each successfully posted invoice, which are propagated back to the account move using the orm
+     * service. Alternatively, the proxy can return an error, in which case the execution of this function should
+     * halt, though the successfully posted invoices are still updated as above.
+     *
+     * @param Array<Object> invoices array representing serialised invoice data and the proxy address to send each invoice to.
+     */
+    async function postInvoices(invoices) {
+
+        // Ping the server to prevent posting to the device when there is no connection to the odoo server
+        try {
+            await rpc("/web/webclient/version_info", {});
+        } catch (e) {
+            // Allow the default error handler to execute after displaying an error message in the dialog
+            state.message = _t("Connection lost, please try again later.");
+            state.error = true;
+            throw e;
+        }
+
+        for (const index in invoices) {
+            let { move_id, messages, proxy_address, company_vat, name } = invoices[index];
+            state.message = sprintf(_t("Posting invoice: %s"), name);
+            try {
+                let deviceResponse = await http.post(
+                    proxy_address + '/hw_proxy/l10n_ke_cu_send', {
+                        messages: messages,
+                        company_vat: company_vat,
+                    },
+                );
+                if (deviceResponse.status === "ok") {
+                    deviceResponse.move_id = move_id;
+                    try {
+                        await orm.call("account.move", "l10n_ke_cu_responses", [[], [deviceResponse]]);
+                        state.successfullySent++;
+                        if (state.successfullySent == invoices.length) {
+                            onAllSent();
+                        }
+                    } catch (e) {
+                        state.message = sprintf(
+                            _t("Error trying to connect to Odoo. Check your internet connection. Error message: %s"),
+                            e.message
+                        );
+                        state.error = true;
+                        break;
+                    }
+                } else {
+                    state.message = sprintf(
+                        _t("Posting the invoice %s has failed with the message: \n %s"),
+                        name,
+                        deviceResponse.state
+                    );
+                    state.error = true;
+                    break;
+                }
+            } catch (e) {
+                state.message = sprintf(
+                    _t("Error trying to connect to the middleware. Is the middleware running? \n Error message: %s"),
+                    e.message
+                );
+                state.error = true;
+                break;
+            }
```

Comment:
All the error processing really makes this code difficult to read.
Here's an option: Keep a variable containing one of the steps you need to take `'1 sendToDevice', '2 responseOK', '3 notifyOdoo'`
raise an error if the device response is not ok
Then use a switch/if in the final catch which sets the error message based on the processing step

Reply:
I'm not sure about this, the conditions don't seem that unreadable to me.
it's just a 
```
try
    if 
        try
        catch
    else
catch
```
Plus I think having the error in situ in the control flow is more readable to me.